src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py

```diff
@@ -0,0 +1,232 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+from collections import defaultdict
+from typing import Dict, List, Set, Union
+
+from snowflake.snowpark._internal.analyzer.select_statement import (
+    Selectable,
+    SelectSnowflakePlan,
+    SelectTableFunction,
+)
+from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
+from snowflake.snowpark._internal.analyzer.snowflake_plan_node import (
+    LogicalPlan,
+    WithQueryBlock,
+)
+from snowflake.snowpark._internal.compiler.query_generator import QueryGenerator
+from snowflake.snowpark._internal.compiler.utils import (
+    replace_child_and_reset_node,
+    reset_node,
+)
+from snowflake.snowpark._internal.utils import (
+    TempObjectType,
+    random_name_for_temp_object,
+)
+
+# common sub-dataframe elimination only operates on top
+TreeNode = Union[SnowflakePlan, Selectable]
+
+
+class CommonSubDataframeElimination:
+    """
+    Optimization that used eliminate duplicated queries that is generated for the
+    common sub-dataframes.
+
+    When the same dataframe is used at multiple places of the plan, the same subquery
+    will be generated at each place where it is used, this lead to repeated evaluation
+    of the same subquery, and causes extra performance overhead. This optimization targets
+    for detecting the common sub-dataframes, and uses CTE to eliminate the repeated
+    subquery generated.
+    For example:
+       df = session.table("test_table")
+       df1 = df1.select("a", "b")
+       df2 = df1.union_all(df1)
+    originally the generated query for df2 is
+        (select "a", "b" from "test_table") union all (select "a", "b" from "test_table")
+    after the optimization, the generated query becomes
+        with temp_cte_xxx as (select "a", "b" from "test_table")
+        (select * from temp_cte_xxx) union all (select * from select * from temp_cte_xxx)
+    """
+
+    # original logical plans to apply the optimization on
+    _logical_plans: List[LogicalPlan]
+    # record the occurrence of each node in a logical plan tree
+    _node_count_map: Dict[TreeNode, int]
+    # record a map between a node and its parents for a logical plan tree
+    _node_parents_map: Dict[TreeNode, Set[TreeNode]]
+    # duplicated node detected for a logical plan
+    _duplicated_nodes: Set[TreeNode]
+    _query_generator: QueryGenerator
+
+    def __init__(
+        self, logical_plans: List[LogicalPlan], query_generator: QueryGenerator
+    ) -> None:
+        self._logical_plans = logical_plans
+        self._query_generator = query_generator
+        self._node_count_map = defaultdict(int)
+        self._node_parents_map = defaultdict(set)
+        self._duplicated_nodes = set()
+
+    def apply(self) -> List[LogicalPlan]:
+        """
+        Applies Common SubDataframe elimination on the set of logical plans one after another.
+
+        Returns
```

Comment:
```suggestion
        Returns:
```
and remove dash lines, and adding indent to the following line, which is consistent inside snowpark python code

Reply:
updated